### PR TITLE
feat(new sink): Initial `signalfx` sink

### PR DESCRIPTION
**Pull request by mengesb**
- Reference: #2562

Co-authored-by: Adrien Pestel <pestouille@gmail.com>
Co-authored-by: Jacob Braaten <jrbraaten6138@gmail.com>
Signed-off-by: Brian Menges <mengesb@gmail.com>

### CODE REVIEW

File: src/sinks/signalfx/metrics.rs [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+
+pub fn default_realm() -> String {
+    String::from("us0")
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone, Default)]
+#[serde(deny_unknown_fields)]
+pub struct SignalFxConfig {
+    // TODO: namespace a du sens dans le contexte SignalFx ? il préfixe les métriques
+    pub namespace: String,
```
Comment (mengesb) [inline src/sinks/signalfx/metrics.rs:173]: https://github.com/timberio/vector/pull/2562#discussion_r423134189
> Is this still to do? Does something need to be resolved before this is ready for release?
> 
> Note that these fields only need to be pub if they are used directly in tests, which does not appear to be the case here.

File: src/sinks/signalfx/metrics.rs [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+        ///*
+        /// Event originated from an agent
+        Agent = 2000000,
+    }
+    //}
+}
+
+pub fn default_realm() -> String {
+    String::from("us0")
+}
```
Comment (mengesb) [inline src/sinks/signalfx/metrics.rs:167]: https://github.com/timberio/vector/pull/2562#discussion_r423134426
> This doesn't need to be `pub`.

File: src/sinks/signalfx/metrics.rs [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+use serde::{Deserialize, Serialize};
+use snafu::{ResultExt, Snafu};
+use std::collections::BTreeMap;
+use std::sync::atomic::{
+    AtomicI64,
+    Ordering::{Acquire, Release},
+};
+use tower::Service;
+
+pub mod signalfx_proto {
```
Comment (mengesb) [inline src/sinks/signalfx/metrics.rs:28]: https://github.com/timberio/vector/pull/2562#discussion_r423134679
> Does this need to be `pub`?

File: src/sinks/signalfx/metrics.rs [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+    // pub mod signalx_proto {
+    //     include!(concat!(
+    //         env!("OUT_DIR"),
+    //         "/com.signalfx.metrics.protobuf.rs"
+    //     ));
+
+    #[derive(Clone, PartialEq, ::prost::Message)]
+    pub struct Datum {
```
Comment (mengesb) [inline src/sinks/signalfx/metrics.rs:36]: https://github.com/timberio/vector/pull/2562#discussion_r423135255
> I think these can all be `pub(super)`. Restricting how public things are helps with dead code elimination.

File: src/sinks/signalfx/metrics.rs [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+    pub realm: String,
+    pub token: String,
+    #[serde(default)]
+    pub batch: BatchEventsConfig,
+    #[serde(default)]
+    pub request: TowerRequestConfig,
+}
+
+struct SignalFxSink {
+    config: SignalFxConfig,
```
Comment (mengesb) [inline src/sinks/signalfx/metrics.rs:184]: https://github.com/timberio/vector/pull/2562#discussion_r423136262
> I don't think you need a copy of the entire config here, so just copy in the required elements individually. In particular, `batch` and `request` (maybe others?) don't appear to be used after the sink is created.

File: src/sinks/signalfx/metrics.rs [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+inventory::submit! {
+    SinkDescription::new::<SignalFxConfig>("signalfx_metrics")
+}
+
+#[typetag::serde(name = "signalfx_metrics")]
+impl SinkConfig for SignalFxConfig {
+    fn build(&self, cx: SinkContext) -> crate::Result<(super::RouterSink, super::Healthcheck)> {
+        let healthcheck = healthcheck(self.clone(), cx.resolver())?;
+
+        let batch = self.batch.unwrap_or(20, 1);
```
Comment (mengesb) [inline src/sinks/signalfx/metrics.rs:205]: https://github.com/timberio/vector/pull/2562#discussion_r423150427
> Is this right? This limits batches to 20 events.

File: src/sinks/signalfx/metrics.rs [hunk 7]
```diff
@@ -0,0 +1,368 @@
...
+    Ordering::{Acquire, Release},
+};
+use tower::Service;
+
+pub mod signalfx_proto {
+    // pub mod signalx_proto {
+    //     include!(concat!(
+    //         env!("OUT_DIR"),
+    //         "/com.signalfx.metrics.protobuf.rs"
+    //     ));
```
Comment (mengesb) [inline src/sinks/signalfx/metrics.rs:33]: https://github.com/timberio/vector/pull/2562#discussion_r423196498
> Can this remnant be removed? It doesn't seem to be adding anything here.

File: src/sinks/signalfx/metrics.rs [hunk 8]
```diff
@@ -0,0 +1,368 @@
...
+    type Output = Vec<Metric>;
+
+    fn encode_event(&self, event: Event) -> Option<Self::Input> {
+        Some(event)
+    }
+
+    fn build_request(&self, events: Self::Output) -> http::Request<Vec<u8>> {
+        let now = Utc::now().timestamp();
+        let interval = now - self.last_sent_timestamp.load(Acquire);
+        self.last_sent_timestamp.store(now, Release);
```
Comment (mengesb) [inline src/sinks/signalfx/metrics.rs:245]: https://github.com/timberio/vector/pull/2562#discussion_r423200547
> This looks like a race condition. Should it use a loop around `compare_and_swap` or `compare_exchange` instead?
> 
> ```rust
> let mut now = Utc::now().timestamp();
> while self.last_sent_timestamp.compare_and_swap(now, AcqRel) != now {
>     now = Utc::now().timestamp();
> }
> ```

File: src/sinks/signalfx/metrics.rs [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+    dpu.encode(&mut buf).unwrap();
+    buf
+}
+
+fn encode_namespace(namespace: &str, name: &str) -> String {
+    if !namespace.is_empty() {
+        format!("{}.{}", namespace, name)
+    } else {
+        name.to_string()
+    }
```
Comment (mengesb) [inline src/sinks/signalfx/metrics.rs:313]: https://github.com/timberio/vector/pull/2562#discussion_r423205223
> I'd prefer a match here, but that's a minor thing.

File: src/sinks/signalfx/metrics.rs [hunk 10]
```diff
@@ -0,0 +1,368 @@
...
+) -> signalfx_proto::DataPointUploadMessage {
+    let series = events
+        .into_iter()
+        .filter_map(|event| {
+            let fullname = encode_namespace(namespace, &event.name);
+            let ts = encode_timestamp(event.timestamp);
+            let dimensions = match event.tags.clone().map(encode_dimensions) {
+                Some(v) => v,
+                None => vec![],
+            };
```
Comment (mengesb) [inline src/sinks/signalfx/metrics.rs:329]: https://github.com/timberio/vector/pull/2562#discussion_r423205879
> I think this `match` could be replaced with `.unwrap_or_default()`.

File: src/sinks/signalfx/metrics.rs [hunk 11]
```diff
@@ -0,0 +1,368 @@
...
+                            int_value: None,
+                        }),
+                        dimensions,
+                    }]),
+                    _ => None,
+                },
+            }
+        })
+        .flatten()
+        .collect::<Vec<_>>();
```
Comment (mengesb) [inline src/sinks/signalfx/metrics.rs:365]: https://github.com/timberio/vector/pull/2562#discussion_r423209672
> It looks like the inner vectors produced by this iterator are all either single elements or `None`. Do I see that right? If so, dropping the inner vector and the trailing `flatten()` seems the best route.

**Comment by mengesb** [comment]
Pulled in comments from #2562 , Can now work through the build failures and comments and perhaps get this implemented

**Comment by binarylogic** [comment]
@mengesb thanks for your contribution! Let us know if you need help to finish this off. Once this PR comes out of draft state we can review it. Otherwise, we'd like to close this if you don't plan on doing that.

**Comment by mengesb** [comment]
Left in draft simply because I wanted to see what state it would be in... and it seems to be in a little bit of a poor state. I might have some time to pick this up... however I'm also having to learn RUST in the process as well... so it's taking a bit of time.

**Comment by binarylogic** [comment]
No problem. If you plan to continue working on it then we'll leave it open. Otherwise, we'd prefer to close it to keep things tidy.

**Comment by binarylogic** [comment]
@mengesb thanks for this contribution! Given that this has gone stale I'm going to close it for now. If you want to pick it back up please feel free. We'd be happy to review and get this integration merged in.

**Comment by NeilJed** [comment]
@mengesb - is this SignalFX still ongoing or have you abandoned it? We have a need for this metrics sink and are happy to try and pick up where you left off if indeed you're no longer proceeding with this.

**Pull request closed without merging**